packages/indexer/src/indexer.test.ts

```diff
@@ -116,4 +133,341 @@ describe("Run Test", () => {
+      }
+
+      if (block.data === "C") {
+        return { filter: { filter: "C" } };
+      }
+
+      return {};
+    };
+
+    const sink = vcr<MockRet>();
+
+    await run(client, indexer, sink);
+
+    // open same db again to check last cursor
+    const db = await open({
+      driver: sqlite3.Database,
+      filename: "file:memdb_indexer?mode=memory&cache=shared",
+    });
+
+    const store = new SqlitePersistence<MockFilter>(db);
+
+    const latest = await store.get();
+
+    expect(latest.cursor?.orderKey).toEqual(108n);
+    expect(latest.filter?.filter).toEqual("BC");
+
+    expect(sink.result).toMatchInlineSnapshot(`
+      [
+        {
+          "data": [
+            {
+              "data": "103B",
+            },
+          ],
+          "endCursor": {
+            "orderKey": 103n,
+          },
+        },
+        {
+          "data": [
+            {
+              "data": "104B",
+            },
+          ],
+          "endCursor": {
+            "orderKey": 104n,
+          },
+        },
+        {
+          "data": [
+            {
+              "data": "105B",
+            },
+          ],
+          "endCursor": {
+            "orderKey": 105n,
+          },
+        },
+        {
+          "data": [
+            {
+              "data": "106BC",
+            },
+          ],
+          "endCursor": {
+            "orderKey": 106n,
+          },
+        },
+        {
+          "data": [
+            {
+              "data": "107BC",
+            },
+          ],
+          "endCursor": {
+            "orderKey": 107n,
+          },
+        },
+        {
+          "data": [
+            {
+              "data": "108BC",
+            },
+          ],
+          "endCursor": {
+            "orderKey": 108n,
+          },
+        },
+      ]
+    `);
+  });
+
+  it("factory mode: last cursor should persist when error is thrown in indexer", async () => {
+    const client = new MockClient<MockFilter, MockBlock>((request, options) => {
+      const [, mainFilter] = request.filter;
+
+      if (Object.keys(mainFilter).length === 0) {
+        expect(request.startingCursor?.orderKey).toEqual(100n);
+
+        return [
+          {
+            _tag: "data",
+            data: {
+              finality: "accepted",
+              cursor: { orderKey: 100n },
+              endCursor: { orderKey: 101n },
+              data: [null, null],
+            },
+          },
+          {
+            _tag: "data",
+            data: {
+              finality: "accepted",
+              cursor: { orderKey: 101n },
+              endCursor: { orderKey: 102n },
+              data: [null, null],
+            },
+          },
+          {
+            _tag: "data",
+            data: {
+              finality: "accepted",
+              cursor: { orderKey: 102n },
+              endCursor: { orderKey: 103n },
+              data: [{ data: "B" }, null],
+            },
+          },
+        ];
```

Comment:
What if you add an `Error` message so we ensure that the indexer `run` function never reaches that block? That way we test if the stop and reconnect logic works exactly we expect.